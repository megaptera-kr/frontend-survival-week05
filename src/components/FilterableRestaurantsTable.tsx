import { useState } from 'react';

import RestaurantsTable from './RestaurantsTable';
import SearchBar from './SearchBar';

import useFetchRestaurants from '../hooks/useFetchRestaurants';
import filterRestaurants from '../utils/filterRestaurants';
import DEFAULT_SELECTED_CATEGORY from '../constants/categories';
import selectCategories from '../utils/selectCategories';

export default function FilterableRestaurantsTable() {
  const [searchKeyword, setSearchKeyword] = useState('');
  const [selectedCategory, setCategory] = useState(DEFAULT_SELECTED_CATEGORY);
  const restaurants = useFetchRestaurants();

  const filteredRestaurants = filterRestaurants(restaurants, {
    searchKeyword,
    selectedCategory,
  });

  const categories = selectCategories(restaurants);

  return (
    <>
      <SearchBar
        value={searchKeyword}
        setValue={setSearchKeyword}
        categories={categories}
        setCategory={setCategory}
      />
      <RestaurantsTable restaurants={filteredRestaurants} />
    </>
  );
}
