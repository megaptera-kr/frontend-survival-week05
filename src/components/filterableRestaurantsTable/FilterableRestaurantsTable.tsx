import { filterRestaurants, getCategories } from '../../utils/restaurant';
import { useFetchRestaurants, useRestaurantsFilter } from './hooks';
import RestaurantsFilter from './restaurantsFilter/RestaurantsFilter';
import RestaurantsTable from './restaurantTable/RestaurantsTable';

export default function FilterableRestaurantsTable() {
  const { restaurants } = useFetchRestaurants();
  const { filter, handleChangeFilter } = useRestaurantsFilter();

  const categories = getCategories(restaurants);
  const filtredCategories = filterRestaurants(restaurants, filter);

  return (
    <div>
      <RestaurantsFilter
        filter={filter}
        categories={categories}
        handleChangeFilter={handleChangeFilter}
      />
      <RestaurantsTable restaurants={filtredCategories} />
    </div>
  );
}
