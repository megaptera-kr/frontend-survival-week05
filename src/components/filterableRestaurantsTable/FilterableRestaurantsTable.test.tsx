import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import FilterableRestaurantsTable from './FilterableRestaurantsTable';

const context = describe;

describe('FilterableRestaurantsTable', () => {
  function renderFilterableRestaurantsTable() {
    return render(<FilterableRestaurantsTable />);
  }

  it('render correctly', async () => {
    const { asFragment } = renderFilterableRestaurantsTable();

    await screen.findByText('메가반점');

    expect(asFragment()).toMatchSnapshot();
  });

  context('when type search text', () => {
    it('restaurants that are included in the search text is rendered', async () => {
      renderFilterableRestaurantsTable();

      const searchInput = screen.getByLabelText('검색');
      await userEvent.type(searchInput, '메가');

      await waitFor(() => expect(screen.getByText('메가반점')).toBeInTheDocument());
    });
  });

  context('when select "한식" category', () => {
    it('only restaurants with "한식" will be rendered', async () => {
      renderFilterableRestaurantsTable();

      const radioBtn = await screen.findByLabelText('한식');
      userEvent.click(radioBtn);

      await waitFor(() => expect(screen.queryByText('흑동고래카레')).toBeNull());
      await waitFor(() => expect(screen.queryByText('메가반점')).toBeNull());
      await waitFor(() => expect(screen.getByText('메리김밥')).toBeInTheDocument());
    });
  });
});
