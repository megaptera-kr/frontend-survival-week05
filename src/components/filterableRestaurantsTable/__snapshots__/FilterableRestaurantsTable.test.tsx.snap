// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableRestaurantsTable render correctly 1`] = `
<DocumentFragment>
  <div>
    <article>
      <h2>
        식당 검색창
      </h2>
      <div>
        <label
          for="search-search"
        >
          검색
        </label>
        <input
          id="search-search"
          name="searchText"
          value=""
        />
      </div>
      <ul
        style="display: flex; flex-direction: row; padding: 0px; gap: 4px; list-style: none;"
      >
        <li>
          <label>
            <input
              checked=""
              name="category"
              type="radio"
              value="전체"
            />
            <span>
              전체
            </span>
          </label>
        </li>
        <li>
          <label>
            <input
              name="category"
              type="radio"
              value="중식"
            />
            <span>
              중식
            </span>
          </label>
        </li>
        <li>
          <label>
            <input
              name="category"
              type="radio"
              value="한식"
            />
            <span>
              한식
            </span>
          </label>
        </li>
        <li>
          <label>
            <input
              name="category"
              type="radio"
              value="일식"
            />
            <span>
              일식
            </span>
          </label>
        </li>
      </ul>
    </article>
    <article>
      <h2>
        식당 목록창
      </h2>
      <table
        style="border: 1px solid black; background-color: black;"
      >
        <caption>
          식당 목록
        </caption>
        <thead
          style="background-color: white;"
        >
          <tr>
            <th
              scope="col"
            >
              업종
            </th>
            <th
              scope="col"
            >
              이름
            </th>
            <th
              scope="col"
            >
              메뉴
            </th>
          </tr>
        </thead>
        <tbody
          style="background-color: white;"
        >
          <tr>
            <td>
              중식
            </td>
            <td>
              메가반점
            </td>
            <td>
              <ul
                style="display: flex; flex-direction: column; padding: 0px; gap: 4px; list-style: none;"
              >
                <li>
                  <div
                    style="display: flex; justify-content: space-between; column-gap: 4px;"
                  >
                    <div
                      style="display: flex; flex-direction: column;"
                    >
                      <span>
                        이름: 짜장면
                      </span>
                      <span>
                        가격: 8,000
                      </span>
                    </div>
                    <button
                      name="#짜장면"
                      type="button"
                    >
                      선택
                    </button>
                  </div>
                </li>
                <li>
                  <div
                    style="display: flex; justify-content: space-between; column-gap: 4px;"
                  >
                    <div
                      style="display: flex; flex-direction: column;"
                    >
                      <span>
                        이름: 짬뽕
                      </span>
                      <span>
                        가격: 5,000
                      </span>
                    </div>
                    <button
                      name="#짬뽕"
                      type="button"
                    >
                      선택
                    </button>
                  </div>
                </li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>
              한식
            </td>
            <td>
              메리김밥
            </td>
            <td>
              <ul
                style="display: flex; flex-direction: column; padding: 0px; gap: 4px; list-style: none;"
              >
                <li>
                  <div
                    style="display: flex; justify-content: space-between; column-gap: 4px;"
                  >
                    <div
                      style="display: flex; flex-direction: column;"
                    >
                      <span>
                        이름: 김밥
                      </span>
                      <span>
                        가격: 3,000
                      </span>
                    </div>
                    <button
                      name="#김밥"
                      type="button"
                    >
                      선택
                    </button>
                  </div>
                </li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>
              일식
            </td>
            <td>
              혹등고래카레
            </td>
            <td>
              <ul
                style="display: flex; flex-direction: column; padding: 0px; gap: 4px; list-style: none;"
              >
                <li>
                  <div
                    style="display: flex; justify-content: space-between; column-gap: 4px;"
                  >
                    <div
                      style="display: flex; flex-direction: column;"
                    >
                      <span>
                        이름: 카레
                      </span>
                      <span>
                        가격: 10,000
                      </span>
                    </div>
                    <button
                      name="#카레"
                      type="button"
                    >
                      선택
                    </button>
                  </div>
                </li>
              </ul>
            </td>
          </tr>
        </tbody>
      </table>
    </article>
  </div>
</DocumentFragment>
`;
